core/src/commonMain/kotlin/com/xebia/functional/xef/llm/assistants/AssistantThread.kt

```diff
@@ -72,208 +69,110 @@ class AssistantThread(
   suspend fun createRun(request: CreateRunRequest): RunObject =
     api.createRun(threadId, request, configure = ::defaultConfig).addMetrics(metric)
 
+  fun createRunStream(assistant: Assistant, request: CreateRunRequest): Flow<RunDelta> = flow {
+    api
+      .createRunStream(threadId, request, configure = ::defaultConfig)
+      .map { RunDelta.fromServerSentEvent(it) }
+      .collect { event ->
+        when (event) {
+          // submit tool outputs and join streams
+          is RunDelta.RunRequiresAction -> {
+            val steps =
+              runSteps(event.run.id).map { metric.assistantCreateRunStep(event.run.id) { it } }
+
+            steps.forEach { step ->
+              val calls = step.stepDetails.toolCalls()
+
+              val run = getRun(event.run.id)
+              if (
+                run.status == RunObject.Status.requires_action &&
+                  run.requiredAction?.type == RunObjectRequiredAction.Type.submit_tool_outputs
+              ) {
+                val callsResult: List<Pair<String, Assistant.Companion.ToolOutput>> =
+                  calls
+                    .filterIsInstance<
+                      RunStepDetailsToolCallsObjectToolCallsInner.CaseRunStepDetailsToolCallsFunctionObject
+                    >()
+                    .parMap { toolCall -> executeToolCall(toolCall, assistant) }
+                    .filterNotNull()
+                val results: Map<String, Assistant.Companion.ToolOutput> = callsResult.toMap()
+                val toolOutputsRequest =
+                  SubmitToolOutputsRunRequest(
+                    toolOutputs =
+                      results.map { (toolCallId, result) ->
```

Comment:
We can potentially also accumulate multiple json parsing errors if you want. I can add it if you like. Something like:

```
listOf("on", "tw", "thre")
    .mapOrAccumulate {
        Either.catch { it.toInt() }.bind()
    } // fold into single Throwable, and rethrow?
```

Reply:
In this case, so far, it's not needed because if a tool fails with an exception, we turn that exception into a json object we send to the LLM to reason about the output.
https://github.com/xebia-functional/xef/blob/dc55080b07f649d11bff74051e62bea7c136dc33/core/src/commonMain/kotlin/com/xebia/functional/xef/llm/assistants/Assistant.kt#L50-L70